### DISCUSSION TITLE
I2S microphone (SPH0645)

### DISCUSSION THREAD

**Discussion by OOHehir** [LIFsCode/ELOC-3.0#24] [General]
Hello all,

I'm trying to integrate the SPH0645 mic into the tflite example [code](https://github.com/OOHehir/tflite-micro-esp-examples/blob/df6497ed8a8eebdfbe707a82beee8b5f31483a60/examples/micro_speech/main/audio_provider.cc#L73) to facilitate testing of models but I'm having some issues getting it working correctly. I see its been implemented into the ELOC code with what appears to be workaround for fixes to the particular mic. Is there any reference for where these fixes came from?

Thanks,

Owen

**Comment by LIFsCode** [comment]
Hi Owen,

unfortunatelly I also have only inherited the Audio part of the ELOC Firmware, so I cannot provice much details about the various workaourds.

But I try to help if I can. Could you be a bit more specifc about what problems occurred?

Just a few comments on what I have seen from the example code so far:

- [code](https://github.com/OOHehir/tflite-micro-esp-examples/blob/df6497ed8a8eebdfbe707a82beee8b5f31483a60/examples/micro_speech/main/audio_provider.cc#L73) uses 16 bit samples, while the [ELOC HW](https://github.com/LIFsCode/ELOC-3.0/blob/master/eloc610LowPowerPartition/src/config.cpp) use 32 bit words and does a configureable gain based bitshift within in [sampler](https://github.com/LIFsCode/ELOC-3.0/blob/master/eloc610LowPowerPartition/lib/audio_input/src/I2SMEMSSampler.cpp) to reduce it to 16 values.
- As far as I understood it, in the end the ELOC should be able to run the tflite model & record data to SD card in parallel, at least as an option (@EDsteve  is that correct). Hence it would make sencs to only have a single I2S Sampling instance which then provices data blocks to the SD card writer and the tflite model. So maybe instead of reading data from i2s directly it could be bettter to have a data block as input. This could also be fed by pre-recorded wave files from an sd card for testing purposes.

@OOHehir if you want we can also make a short call (signal, skype, whatever) to discuss the system structure

Fabian

**Reply by EDsteve** [reply]
Correct. Once an elephant sound has been detected by the model. The sound should be recorded to the SD card plus a few more seconds if possible.

**Comment by EDsteve** [comment]
@OOHehir 
Not sure if that info helps. But our I2S code is based on [atomic14 github.](https://github.com/atomic14/esp32_sdcard_audio) 
You can just try to make his code work first. And go from there. That should be more easy.
An even more simple I2S test code (without SD card) is [this one here](https://github.com/atomic14/esp32-i2s-mic-test) also from atomic14.

TIP: If you get no sound at all. It's worth playing with L/R pin of the microphone. Meaning that you should always try to connect L/R to GND or VCC.

**Comment by OOHehir** [comment]
Thanks both for the responses.

At the moment I'm trying to get a speech detection TensorFlow micro (TFM) example working as a standalone application. I think going this route would be useful to isolate an issues that may come up when integrating it with the main code-base. 

The issue is that the TFM example uses a different mic & trying to find a reliable example of code that works for this mic is proving tricky. I see examples that use 16 bit settings, some 32 bit, some make reference to only working on one channel (I think that fits your comment ED about L/R pin), some that use bit shifting on the samples received, some state a minimum sample rate of X is required, etc etc.. 

The application compiles & runs OK & in certain mic configurations appears to receive sound input but only 'silence' or 'unknown', I suspect because the mic configuration is incorrect.

I'll continue plugging away & see if I can get it working!

**Comment by LIFsCode** [comment]
If you are actually using a SPH0645 microphone, the 32 bit is **mandatory**. It requires an internal oversampling factor of 64, i.e. 64 clock pulses per WS period.
See [datasheet](https://www.mouser.de/datasheet/2/218/ph0645lm4h_b_datasheet_rev_c-1525723.pdf) page 6 for details.

But better double check the code marking on the IC, as the cheap breakout boards sometimes differ in ICs whihc are actually soldered.

Concerning debugging:

- Are you sure your mic is working? I have a broken mic module, which outputs only light noise around flat line. Probably it got too hot during soldering. Double check with a basic example
- I would suggest running the simple [atomic14 github.](https://github.com/atomic14/esp32_sdcard_audio) code and have a look at the wavefiles created that way, or simple print out the values to get an idea if it's flatline (most likely wrong WS, i.e. left/right channel), or low amplitude noise (potentially wrong bit selection or broken mic).
- I noticed that at some point within the espressif framework development the channel selection changed. I had tests with arduino based code (running ESP IDF 4.x ) working with right channel only, while pure IDF based code (running ESP IDF 6.x) works with left channel only, without any changes in HW setup. Since Arduino framework is always a bit behind its limited to use with IDF 4.x as base.

**Comment by OOHehir** [comment]
Thanks for that, I see also in the datasheet data precision is 18 bits so that's where the 14 bit shifting I see is coming from..

Will try the code & let you know..

**Comment by OOHehir** [comment]
Well a bit of progress..

I put a scope on the BCLK & DOUT lines of the mic & can see the waveforms change in response to audio input. I can also see some values change in the buffer that's obtained in response to noise.

It seems that the missing element maybe the [bit-shifting](https://github.com/LIFsCode/ELOC-3.0/blob/3e34414f753c7800f2e8c24105a0be8899070756/eloc610LowPowerPartition/lib/audio_input/src/I2SMEMSSampler.cpp#L54) that's required...

**Comment by OOHehir** [comment]
Finally some headway! I managed to get the [example](https://github.com/atomic14/esp32_sdcard_audio) working as suggested. Initially using the code from the original project the quality was fairly poor, lots of scratching & unintelligible sounds. However amending the code as per the [ELOC code](https://github.com/LIFsCode/ELOC-3.0/blob/3e34414f753c7800f2e8c24105a0be8899070756/eloc610LowPowerPartition/lib/audio_input/src/I2SMEMSSampler.cpp#L53) resulted in a significant improvement. 

Now I know that the mic setup is working & what code to use hopefully I'm make some further progress on the TFLite example..

**Comment by OOHehir** [comment]
Well its taken a while but I believe I finally got my head around I2S mics & sound sampling but please correct me if the following is wrong....

I've been trying to work out the snippet of code that appears [here](https://github.com/LIFsCode/ELOC-3.0/blob/ffbfa34a93b67261220cdbeb4eb522a036f60025/eloc610LowPowerPartition/lib/audio_input/src/I2SMEMSSampler.cpp#L54), i.e:

`samples[i] = graw_samples[temp++]  >> mBitShift; //14 mahout, 11 forest`

There are some comments about different shift values, some which work, some which don't. They seemed to be 'magic' values without explanation. What I think is correct is that previously, at some point, a 16 bit mic was used. As detailed [here](https://github.com/LIFsCode/ELOC-3.0/blob/6116e98e0d74c040b3967335faff0129ef878fa8/eloc610LowPowerPartition/lib/audio_input/src/I2SMEMSSampler.cpp#L199) the data arrives from the DMA MSB first. So if a 16 bit value arrives into a 16 bit datatype everything is aligned. However for the ELOC 3.2 board the mic is 24bit & a 32bit data type is used to store it , so this data must be shifted 8 bytes to the right to correctly align it. The resulting sample is probably too quiet so must be increased. An efficient way of achieving this is by shifting data to the left. Every left shift doubles the volume. The most efficient scheme is therefore to work out the overall shift required & accomplish it in one operation, rather than right shifts followed by left shifts. The 'processed' sample is then send to the wav file which, stores data in 16 bit format. 

I noticed during experimentation that if a particularly loud sound was detected it may exceed the limit of 16 bits, causing particularly unpleasant distortion & making the portion of the sample unintelligible. I therefore implemented a [method](https://github.com/LIFsCode/ELOC-3.0/blob/6116e98e0d74c040b3967335faff0129ef878fa8/eloc610LowPowerPartition/lib/audio_input/src/I2SMEMSSampler.cpp#L227) of detecting these & clipping the value to its limit, greatly reducing the distortion.

Finally, something which may or may not be useful, I introduced a method of visualising the sampled sound from the board that can be enabled/ disabled in a simple manner. Its detailed [here](https://github.com/LIFsCode/ELOC-3.0/blob/84a4df9bfd800c073e5703382e473641a7a395f8/eloc610LowPowerPartition/lib/audio_input/src/I2SMEMSSampler.cpp#L113). It may prove useful for development purposes. Using it its possible to get images of the sound such as that below.

From the discussion above its seems to me that removing the mBitShift configuration value & replacing it with a simple volume factor (i.e. x 1, x 2, x 4, x 8, x 16) may be clearer approach. 

![Screenshot from 2023-11-08 22-32-16](https://github.com/LIFsCode/ELOC-3.0/assets/40183643/f21e8dac-0752-42c8-b662-4607e7d5214c)

**Reply by EDsteve** [reply]
@OOHehir 
That sounds right to me. I don't understand the code 100%. But i understood the reason behind mBitShift. And that you made it much better now.  :)
I guess i could have made your life easier with an explanation about mBitShift earlier. 

Tom, who wrote the first ELOC firmware two years ago, introduced mBitShift for the same reason as you desribed already: The problem of 24bit microphones and 32bit data type. 
So he shifted the data 8 bits to make it work. Which at the same time made it possible to set microphone gain. As you already explained earlier. Shifting it left or right increases or decreaes the gain (volume). So in the ELOC app we can set the gain by changing these numbers (14 mahout, 11 forest). 14 = low gain and 11(more quiet) = high gain (louder)

But in my understanding it comes with a huge downside: We loose around 30% of the microphone's dynamic range.
If the mBitShift is set to 11 (high gain). Then we can hear distant (quiet) sounds more clear. But if an elephant trumpets very close to the ELOC. Then the sound will be distorded. Which results in pure noise recordings.
The dynamic range of the microphone is good enough to handly both scenarios. But because of the bitshift we kind of cripple the microphones dynamic rancge drastically. I assume there is no way around it.

So if i understand your solution correctly:
If an elephant trumpet is very close. We would be still able to hear the trumpet (instead of total noise). But with missing information?
We are still able to set gain using the volume factor?

Thanks for the upgrad. That seems very promising.
Cheeers
ED

**Comment by LIFsCode** [comment]
@OOHehir I just came to an idea to get rid of the bit shifting and thus rid of the clipping or noise distortion effects. I'm not too deep into the AI stuff, so please correct me if I'm wrong with some assumptions:

### AI Detetion
- As far as I know the EI SDK runs on floats internally, and the input is first converted from int  (16bit) to float using numpy::int16_to_float. But the numpy lib  also support 32bit to float conversion using numpy::int32_to_float. Would it be an option to modify the I2S Sampler to output the full 32 bit values (with 24 significant bits from the mic), run the EI based on the 32 bit value and leave it to the SD card writer to handle the bit shift
- **Problem**: I have no idea how this affects the precision of the model if it is not trained with 24 bit input data? But could be worth a try

### SD Card Recording
- For the recording the compression to 16 bit makes sense in terms of SD card usage. @EDsteve this is question is up to you: how limiting it the size of the sd card? The wave file is capable of storing 32 bit as well, but leading to a 50% reduction of total recording time as a downside, while improving the dynamic range
- If continous 32 bit storage is not an option: Now that we have the extended PSRAM it would be possible to modify the WAVWriter to not directly store the data on SD card, but:
1. Store the full 32 bit as received from the MIC in an internal buffer (limited by max. allowed seconds per file (@EDsteve : do we have a limit here? or may we introduce a limit?)
2. Check the max. amplitude within this data and set the volume scaling factor (MIC bitshift) accordingly.
3. Compress the data to 16 bit with the dynamic MIC bitshift (2) and store the MIC bitshift value with this file for postprocessing
4. In a post processing (at a PC) the different files have to be normalized, as they may have different volume levels. This can be done offline once, the data is read from the SD card

- This would allow to store always the optimal dynamic range, However efficiency depends highly on the length (in seconds) of a file. The longer the recording file, the worse the efficiency.

**Comment by OOHehir** [comment]
@LIFsCode Yes, I think it would be fairly straightforward to pass 32 bit values into the model, rather than 16 bit values. It would needed to be confirmed that this wouldn't effect the result of the EI model but I don't think it would.

@EDsteve @LIFsCode I think it may be possible to store the pulse-code-modulation data (PCM), contained in a wav file, at 24 bits rather than 16 bit. This would increase the quality while helping to conserve space (as opposed to 32 bit data). However CD quality is 44.1KHz at 16 bit so I wonder if the additional storage requirement would be beneficial... 

**Proposed solution**
I came up with an idea which I was going to propose some days ago which actually may solve this issue.....  It basically boils down to automatic volume gain adjustment.. It would work something like:

- Volume is adjusted in the values 2 (min), 4, 8 or 16 (max)
- The samples are continuously checked & if, say over a period of 30 secs, a significant number of samples don't reach at least, say 50% of the max range the volume is increased (doubled). Volume is only increased every 30 sec
- If a significant number of samples exceed 95% of the max range the volume is reduced (halved) at next sampling
- This scheme would allow a slow ramp up of volume but a quick reaction to excessive volume
- At all times the samples will be clipped if they exceed the max range

The downside of this automatic adjustment is that it won't be possible to know from the recording how far the object was from the sensor.

If accepted the Bluetooth interface would be modified to:
Volume adjustment: Manual/ Automatic
Volume level: 2, 4, 8, 16 (used to set manual level or starting point for automatic)


Edit:
Just thought of a further improvement/ modification:
Volume adjustment only happens at the start of wav file recording & the volume level is appended to the filename.

**Comment by EDsteve** [comment]
@LIFsCode @OOHehir 
I didn't expet to see so many solutions :) 

**Storing wave files as 32bit**
Recording-wise a good solution. But that would increase power consumption and as already mentioned reduces the possible recording time by 50% (on the same sized SD card) -> Big deal.
We would also have double the amount of data. Which is another challange. One hour recording = 200MB. Most projects use 50 recorders running 24/7 for a few months.

**Automatic gain control**
I think both of you suggested this solution in different ways. And that sounds definitely like a good option to have. Of course optional in the app. This solution would give every file a different "volume/gain". But as long the user can choose. It's fine.
Post processing can fix this. But to copy and upload all the files is already a too big challange for many biologist. Haha. Doing post processing is for sure one step too much. Biologist want to deal as less as possible with technical things.

We have used the code with the problematic dynamic range for two years now and in my opinion it works okay enough. The chances that an elephant is that close to the recorder is small. Wind is a bit more problematic and makes the recording "clip" easily. But with strong wind. We don't hear/see much of other sounds in the recording anyway. 
If that takes too much time to solve. Maybe it's best to work on other issues and come back to it later? We can use it as is for now. Having the AI and LoRa working is still higher in the priority list. (Words from above) Haha


Cheeers
ED

P.S.: The user must be able to choose 3600 for the max seconds per file. That would make the buffer blow up i guess :)

**Comment by LIFsCode** [comment]
@EDsteve Makes sense. Also from a quality point of view I agree with @OOHehir this would be a waste I think it's far from having a real quality benefit.
So my prefered option would be keep with the 16 bit fixed volume shift for the recording. Probably if there is so much noise distortion due ot wind/rain it wouldn't be useable for biologist anyway

But for the AI I detection I think it would be usefull to switch to 32 bit to float conversion, as long as it does not affect the model. I could immage that the detection logic could still work even with high levels of noise from wind/rain simple because they will differ from the frequency characteristic. So here it could make a real difference if we could have the full dynamic range available.

**Comment by OOHehir** [comment]
@EDsteve @LIFsCode 

So just to make sure we agree on where we're going:

I''l modify the ai/ei buffers to 32 bits & feed the model with 32bit -> float conversion. This should be quick to achieve.

It would seem that the following options are non-runners for the wav file:
  - post-processing of the file, difficult to achieve for a 3600sec recording
  - using a single fixed 'automatic' gain set at the start of the recording. This would be as a result of the scenario where the gain started high & partway through a 3600sec recording a loud noise source came in close proximity resulting in a significantly distorted recording.
  
So @EDsteve would you think a gain that changed during the recording be useful? I could probably put the code together reasonably quickly but there would be no way to tell from the wav file after recording what the gain was used at any point. If you think it's more beneficial for me to work on some other aspect/ issue that's fine. 

Regardless, hopefully by introducing the clipping mechanism the wind distortion issue will be addressed somewhat.

**Comment by EDsteve** [comment]
@OOHehir @LIFsCode 

Using 32bit instead of 16bit for the AI. Anybody an idea how much power it will draw?
I would like to compare if or what difference it makes for the AI. So if it is a quick job. I would prefer to be able to compare both. 16bit and 32bit with our model(s).

Our recorders will be used for different wildlife puposes next year. I am sure a automatic gain solution will be beneficial in some projects. We are testing some ELOCs in SriLankga right now where wind is a big problem. Might come handy in there.
If it's a quick thing. Yes. Go for it :)
But as mentioned earlier. Getting the AI running is still the number one priority. I still owe you an email regarding my testes. Coming soon. I just always run out of time every day. Haha

**Comment by OOHehir** [comment]
Closed as current code has working solution
